Octoprint 1.4 compatibility [5]

amrsoll opened the pull request:
### Main Changes : 

- Refactoring, updating dependencies
- Octoprint hook factories in separate `hooks.py` file or other relevant files (previously `__init__.py`)
  - new hook : loginui theming for the login screen presented before loading all of the front-end
- Disable new builtin plugins 
- Major refactoring of `comm_acc2.py`
  - Inherits from relevant classes written in `octoprint.util.comm`
  - LOCKED state is now part of the OPERATIONAL group of states
  - Reference as much hardcoded info from `octoprint.util.comm` as possible
  - Increased GRBL read window (timeout)
  - Close connection more gracefully
  - Except for the "redo job" functionality, all of the gcode file handling code is refering to the OP implementation.
- Remove the hacks that were replace with hook options
  - no monkey patching of the analysis queues for filemanager handling.
- `@logExceptions` on functions called from the front-end
- Fix bug when area dimension of workspace not communicated to front end
- Change dependencies around for `setup.py`

### Testing

1. Back up the virtualenv `cp -r ~/oprint ~/oprint.back` (Reason : I've had trouble reverting the 1.4 update)
2. Activate virtualenv `. ~/oprint/bin/activate` and `pip install --upgrade pip==18.0` - **NOT** higher than 18 or it won't run with python2
3. `pip  install --upgrade setuptools` and `sudo apt install python-wrapt`
4. Either:
    - Copy over the source code from the mrbeam/OctoPrint `f_octoprint_1.4` branch and `pip install ./OctoPrint`
      Or
    - Switch the software update branch in `config.yaml` to track the `f_octoprint_1.4` branch and perform the update through the gui
5. rsync or change software update tracking branch for MrBeamPlugin -> branch `f_octoprint_1.4_rebased`
6. Finally `restart_octoprint`

### TODO list 
- Migration
  - [ ] `pip install --upgrade pip==18.0`
  - [ ] [disable safe_mode](https://docs.octoprint.org/en/1.4.2/features/safemode.html)
  ```yaml
  # .octoprint/config.yaml
  server:
    ignoreIncompleteStartup: true
  ```

- [x] Change Startup screen - needs to allow Access Management and setup login wizard.
      Alternatively, the user has limited rights (such as not being able to install extra plugins)
- [ ] Change the UI plugin template to add a logo, ...
- [x] HowTo / Test retrograde to 1.3.6 through software update plugin.
- [ ] Try 1.5 out first
- [ ] Write down new features we can leverage
- [ ] GRBL
  - [ ] Flashing
  - [ ] Reset

Sorry @irlaec the rebased commit messages didn't all mention you :open_mouth:

irlaec committed:
Fix backend issues: printer, comm_acc, events

amrsoll committed:
wip: change usersVM, fix deprecated jquery

amrsoll committed:
Change deprecated flask babel import

amrsoll committed:
Hide unnecessary settings

amrsoll committed:
Set py compatibility + change setup.py

amrsoll committed:
Analysis Queue refactor for OPv1.4

amrsoll commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Octoprint 1.4](https://trello.com/c/pGFNsysf/371-octoprint-14)

amrsoll committed:
Fix deprecation warnings.

amrsoll committed:
Light refactor of the Printer class.

amrsoll committed:
Larger refactor of comm_acc.

Uses more of the OP functions to process Gcode.

amrsoll committed:
Fix conflicting MachineCom states + timeout

- Settings timeout increased for slow I/O
- Minor refactors

amrsoll committed:
Locked State is now in Operational.

And log exceptions when they happen in sendCommand.

amrsoll committed:
Fix homing and _comm command kwargs.

amrsoll commented:
Current problem:
MotherViewModel produces a 500 Error when it queries for the sliced result.
https://github.com/mrbeam/MrBeamPlugin/blob/9c300c563ffd58dc44fb3ab7dc35116cf029dfaa/octoprint_mrbeam/static/js/mother_viewmodel.js#L655
```
self.gcodefiles.loadFile(data, false); // FIXME Err 500 - loads gcode into gcode viewer
```

amrsoll commented:
The login screen also needs that spicy Mr Beam look.

This page shows before anything else gets loaded, saving cpu resources if unauthorised people access the page.

![Screenshot from 2020-10-27 11-05-08](https://user-images.githubusercontent.com/24782867/97286978-53efcf80-1844-11eb-8e8d-1584946d706f.png)

amrsoll commented:
New Keys in config files

`.octoprint/config.yaml`

```yaml
accessControl:
    salt: y1aF [...] E23u
api:
    key: 900B [...] D3B6

feature:
    modelSizeDetection: false
    printCancelConfirmation: false
    sdSupport: false

    errortracking:
        unique_id: eb1d567b-1422-4056-9168-34300c42dd00
plugins:
    tracking:
        enabled: true
        unique_id: 3a665 [...] ff07cb6e3409
server:
    ignoreIncompleteStartup: true
    pluginBlacklist:
        enabled: true
    secretKey: 7B2qjZLr1Ub [...] KPHTmYHsio
```

`users.yaml`
```yaml
this@user.com:
  apikey: D6F9B33 [...] 5A3950D4
    groups:
    - admins
    - users
    permissions: []
```

amrsoll committed:
Inherit MachineCom and log some exceptions.

amrsoll committed:
Fix state handling (getStateId) & minor regression.

amrsoll committed:
Change getStateId for better child STATE handling

amrsoll committed:
Can now laser normal jobs - Various fixes.

ghost commented:
## DeepCode's analysis on [#eb8dae](https://www.deepcode.ai/app/gh/mrbeam/MrBeamPlugin/50f383df745ffa9372a28895fffc97b0acb0cc1a/mrbeam/MrBeamPlugin/eb8dae7b0f7bc868fc385245c178bc120dc6dc61/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=4&) found:

- :information_source: **4** minor issues. :point_down:
- :heavy_check_mark: **4** issues were fixed.
## Top issues

<table>
  <thead>
    <tr>
      <th align="left">Description</th>
      <th align="left">Example fixes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td width="77%">Either all return statements in a function should return an expression, or none of them should. Occurrences: <ul><li><a href="https://github.com/mrbeam/MrBeamPlugin/blob/eb8dae7b0f7bc868fc385245c178bc120dc6dc61/octoprint_mrbeam/hooks.py#L35">hooks.py:35</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/mrbeam/MrBeamPlugin/50f383df745ffa9372a28895fffc97b0acb0cc1a/mrbeam/MrBeamPlugin/eb8dae7b0f7bc868fc385245c178bc120dc6dc61/pr/_/%2Foctoprint_mrbeam%2Fhooks.py/python%2Fpylint%2FR1710/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
    <tr>
      <td width="77%">Import "from octoprint_mrbeam.__version import __version__" should be placed at the top of the module Occurrences: <ul><li><a href="https://github.com/mrbeam/MrBeamPlugin/blob/eb8dae7b0f7bc868fc385245c178bc120dc6dc61/octoprint_mrbeam/__init__.py#L34">__init__.py:34</a></li> <li><a href="https://github.com/mrbeam/MrBeamPlugin/blob/eb8dae7b0f7bc868fc385245c178bc120dc6dc61/octoprint_mrbeam/__init__.py#L35">__init__.py:35</a></li> <li><a href="https://github.com/mrbeam/MrBeamPlugin/blob/eb8dae7b0f7bc868fc385245c178bc120dc6dc61/octoprint_mrbeam/__init__.py#L36">__init__.py:36</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/mrbeam/MrBeamPlugin/50f383df745ffa9372a28895fffc97b0acb0cc1a/mrbeam/MrBeamPlugin/eb8dae7b0f7bc868fc385245c178bc120dc6dc61/pr/_/%2Foctoprint_mrbeam%2F__init__.py/python%2Fpylint%2FC0413/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
    <tr>
      <td width="77%">Unused logme imported from octoprint_mrbeam.util.log Occurrences: <ul><li><a href="https://github.com/mrbeam/MrBeamPlugin/blob/eb8dae7b0f7bc868fc385245c178bc120dc6dc61/octoprint_mrbeam/printing/printer.py#L6">printer.py:6</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/mrbeam/MrBeamPlugin/50f383df745ffa9372a28895fffc97b0acb0cc1a/mrbeam/MrBeamPlugin/eb8dae7b0f7bc868fc385245c178bc120dc6dc61/pr/_/%2Foctoprint_mrbeam%2Fprinting%2Fprinter.py/python%2Fpylint%2FW0611/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
  </tbody>
</table>


#### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/mrbeam/MrBeamPlugin/50f383df745ffa9372a28895fffc97b0acb0cc1a/mrbeam/MrBeamPlugin/eb8dae7b0f7bc868fc385245c178bc120dc6dc61/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=4&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=mrbeam)

amrsoll committed:
Fix a case w/ undefined var (should refactor?)

amrsoll committed:
Fix state flags (adds locked state)

amrsoll committed:
Fix lasering gcode.

amrsoll commented:
### TODO
- [ ] Make a custom login screen
- [ ] ~Make a custom safe_mode / fallback screen~
- [x] Remove the extra `Update` buttons on the software update screen
- [x] Welcome dialog -> Enable Access management by default

![Screenshot from 2020-12-02 17-49-50](https://user-images.githubusercontent.com/24782867/100929321-bdb86480-34e7-11eb-80c9-4a30dd1662a1.png)

amrsoll committed:
Disable individual `Update` buttons.

amrsoll committed:
rm log before merge (simplifies merge)

amrsoll committed:
Merge branch 'develop' into f_octoprint_1.4_rebased

amrsoll committed:
Comment @logExceptions while try / except present.

amrsoll commented:
Options to modify the loginui template :
- Create new (mrb_loginui) plugin, new package and inherit the `LoginUiPlugin` -> modify 2 functions
- monkey-patch the loginui plugin to use a different template file.
- Simply change the css - styling : button, background...
- Make an upstream change to the loginui that allows to change the template through a hook, and wait for the following OP release.


![Screenshot from 2020-12-02 23-56-35](https://user-images.githubusercontent.com/24782867/100941515-0aa53680-34fa-11eb-8b18-07acb104e693.png)

amrsoll commented:
For the safe_mode / fallback screen : Not urgent / needs planning / bigger (out of scope) task

amrsoll committed:
Small Refactor and fixes.

amrsoll committed:
Remove debug logging lvl

amrsoll committed:
movehooks to separate file + login theming.

amrsoll commented:
Login theming requires a fix for mrbeam/Octoprint 

2 problems when adding just a css theme: 
- We cannot override the name in the tab
- We cannot change the href of the password recovery link

![Screenshot from 2020-12-03 11-40-06](https://user-images.githubusercontent.com/24782867/100998934-6ef8e200-355c-11eb-8d21-cee55eb35926.png)

amrsoll committed:
Small wizard refactor.

amrsoll commented on octoprint_mrbeam/led_events.py:
```diff
@@ -192,9 +192,9 @@ def _execute_command(self, command, commandType, debug, event=None, payload=None
+					processedCommand.append(self._processCommand(c, event, payload))
             else:
-                processedCommand = self._processCommand(command, payload)
+				processedCommand = self._processCommand(command, event, payload)
```
careful, check indentation

amrsoll committed:
Reduce logs & fix deprecation.

amrsoll committed:
More deprecation fixes.

amrsoll committed:
Fix missing import.

amrsoll committed:
Fix indentation (tab - spaces).

amrsoll committed:
User created in firstrun is Admin.

khaledsherkawi commented:
Final OctoPrint 1.4 Testing notes:
- Pressing the laser button on the device while the job is being processed DOES NOT pause the job
- Pausing the job from the UI works but cancelling the job DOES NOT work
- Upon pausing, one CAN NOT resume the job in any way (neither via pressing button on the device nor via clicking on the "press the button to continue" text in the DEV ENV)
- Opening the lid DOES NOT stop the lasering - it is not detected at all by the UI
- Upon opening the lid, the camera DOES NOT reflect any changes to the working area (might not be related to OctoPrint 1.4 according to @amrsoll)
- Upon logging out, the user see a different screen to fill in their credentials rather than the usual Modal window shown. Only after they fill their credentials and log in, the Mr. beam plugin loading will start (unlike before)
- There is an extra "Application Keys" tab in the "User Settings" modal window
- There are 3 extra tabs (Online connectivity check, Plugin blacklist, CuraEngine) in the Welcome "Dialog modal" upon first run
- Settings -> Access control: Text missing + Extra "Groups" tab + "Admin" option missing + "Add User" option has much more options (permissions + groups) + minor styling modifications
- Settings -> Logging: Extra Logging Level section + different styling
- Extra "CuraEngine" tab in "Settings"

Please find the detailed test results attached below.
[Regression Testing - OctoPrint 1.4.xlsx](https://github.com/mrbeam/MrBeamPlugin/files/5815327/Regression.Testing.-.OctoPrint.1.4.xlsx)

Please note that further testing is needed once the issues in the notes above are resolved

amrsoll committed:
Merge branch 'develop' into f_octoprint_1.4_rebased

amrsoll committed:
Comm layer fixes

- Fix pause / Resume when pressing the button
- Fix duplicate events triggered
- Remove support for arduino
- Fix LED eventCallback error

amrsoll committed:
Multiple bugfixes

- Replace deprecated jquery ``onload``
- Fix unicode errors in mrb_logger
- Fix crash in analytics because of wrongful ``get()``

amrsoll committed:
Minor GRBL flash refactor

- Removes support for on old version of GRBL from 2017

amrsoll committed:
Refactor LED events when using system commands.

JFSolution commented:
tested on latest commit see comment in jira issue https://mr-beam.atlassian.net/browse/SW-70?focusedCommentId=10212

amrsoll committed:
Bump version.

amrsoll committed:
Disable Safe Mode through migration.

khaledsherkawi committed:
active class added only to first element (#1310)

amrsoll committed:
Fix file selection events

- Only fire FILE_SELECTED event once (done by OctoPrint)
- Use `printer.start_print()` instead of `printer.select_file(printAfterSelect=True)`

JFSolution commented on octoprint_mrbeam/__init__.py:
```diff
@@ -3071,7 +3014,18 @@ def __plugin_load__():
                 disabled=dict(
                     wizard=["plugin_softwareupdate"],
-                    settings=["serial", "webcam", "terminalfilters"],
+                    settings=[
```
we have to check which of these we can move to the disabled plugins section

JFSolution commented on octoprint_mrbeam/iobeam/onebutton_handler.py:
```diff
@@ -166,7 +172,7 @@ def onEvent(self, event, payload):
                 self.print_started > 0
                 and time.time() - self.print_started > 1
-                and self._printer.get_state_id() == self.PRINTER_STATE_PAUSED
+                and self._printer._comm._state == self.PRINTER_STATE_PAUSED
```
can be changed to self._printer.is_paused()

JFSolution commented on octoprint_mrbeam/led_events.py:
```diff
@@ -133,29 +133,23 @@ def __init__(self, plugin):
-            commandType = "system"
-            debug = False
-            self._execute_command(command, commandType, debug)
+        if isinstance(fps, int) and 15 <= fps <= 45:
```
these values shouldn't be hardcoded

JFSolution commented on octoprint_mrbeam/migrate.py:
```diff
@@ -213,6 +214,15 @@ def run(self):
                     equal_ok=False,
                 ):
                     self.hostname_helper_scripts()
+                if self.version_previous is None or self._compare_versions(
```
maybe do this directly in octoprint settings as default settings

JFSolution commented on octoprint_mrbeam/printing/comm_acc2.py:
```diff
@@ -2992,52 +2819,43 @@ def _calc_total_lines(self):
 
 
 class PrintingGcodeFromMemoryInformation(PrintingGcodeFileInformation):
+    # FIXME gcode is a string, not a file or stringIO (cannot be opened or interpreted as bytes)
```
have to be checked, if the corner calibration lasering is still working

JFSolution replied:
it seems like it rembembers the old material settings
